File: lib/python/opentoken/src/main/opentoken/attributes/person/first_name_attribute.py
from typing import List
import re
from opentoken.attributes.base_attribute import BaseAttribute
from opentoken.attributes.utilities.attribute_utilities import AttributeUtilities
from opentoken.attributes.validation.not_in_validator import NotInValidator


class FirstNameAttribute(BaseAttribute):
    """Represents the first name of a person.

    This class extends BaseAttribute and provides functionality for working with
    first name fields. It recognizes "FirstName" and "GivenName" as valid aliases
    for this attribute type.

    The attribute performs no normalization on input values, returning them
    unchanged.
    """

    NAME = "FirstName"
    ALIASES = [NAME, "GivenName"]

    # Pattern to match and remove common titles
    TITLE_PATTERN = re.compile(
        r"(?i)^\s*(mr|mrs|ms|miss|dr|prof|capt|sir|col|gen|cmdr|lt|"
        r"rabbi|father|brother|sister|hon|honorable|reverend|rev|doctor)\.?\s+",
        re.IGNORECASE
    )

    # Pattern to match trailing periods and middle initials in names.
    #
    # This pattern matches:
    #  - A space, followed by a single non-space character (middle initial),
    #    optionally followed by a period, at the end of the string.
    #
    # Breakdown of the regex:
    #   \s           A space
    #   [^\s]        Any single non-space character (middle initial)
    #   \.?          Optional period
    #   $            End of string
    TRAILING_PERIOD_AND_INITIAL_PATTERN = re.compile(r'\s[^\s]\.?$')

    def __init__(self):
        placeholder_values = AttributeUtilities.COMMON_PLACEHOLDER_NAMES
        validation_rules = [NotInValidator(placeholder_values)]
        super().__init__(validation_rules)

    def get_name(self) -> str:
        return self.NAME

    def get_aliases(self) -> List[str]:
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Returns the value unchanged after removing titles."""
        if not value:
            return value

        normalized = AttributeUtilities.normalize_diacritics(value)

        without_title = re.sub(self.TITLE_PATTERN, '', normalized).strip()

        if without_title:
            normalized = without_title

        without_suffix = AttributeUtilities.remove_generational_suffix(normalized)

        if without_suffix:
            normalized = without_suffix

        normalized = re.sub(self.TRAILING_PERIOD_AND_INITIAL_PATTERN, '', normalized).strip()

        # Remove non-alphabetic characters
        normalized = AttributeUtilities.NON_ALPHABETIC_PATTERN.sub('', normalized)

        # Normalize whitespace
        normalized = AttributeUtilities.WHITESPACE_PATTERN.sub(' ', normalized)

        return normalized

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/person/FirstNameAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import java.util.List;
import java.util.regex.Pattern;

import com.truveta.opentoken.attributes.BaseAttribute;
import com.truveta.opentoken.attributes.utilities.AttributeUtilities;
import com.truveta.opentoken.attributes.validation.NotInValidator;

/**
 * Represents the first name of a person.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * first name fields. It recognizes "FirstName" and "GivenName" as valid aliases
 * for this attribute type.
 * 
 * The attribute performs no normalization on input values, returning them
 * unchanged.
 */
public class FirstNameAttribute extends BaseAttribute {

    private static final String NAME = "FirstName";
    private static final String[] ALIASES = new String[] { NAME, "GivenName" };

    /**
     * Pattern to match and remove common titles and title abbreviations from first names.
     *
     * This pattern matches (case-insensitive):
     *  - Common titles such as "Mr", "Mrs", "Ms", "Miss", "Dr", "Prof", etc.
     *  - Optional period after the title (e.g., "Dr.")
     *  - One or more spaces following the title
     *
     * Breakdown of the regex:
     *   (?i)                         Case-insensitive flag
     *   ^                            Start of string
     *   (mr|mrs|ms|miss|dr|prof|     List of supported titles
     *    capt|sir|col|gen|cmdr|lt|   More supported titles
     *    rabbi|father|brother|sister|
     *    hon|honorable|reverend|rev|doctor)
     *   \.?                         Optional period
     *   \s+                         One or more spaces after the title
     */
    private static final Pattern TITLE_PATTERN = Pattern.compile(
            "(?i)^(mr|mrs|ms|miss|dr|prof|capt|sir|col|gen|cmdr|lt|rabbi|father|brother|sister|hon|honorable|reverend|rev|doctor)\\.?\\s+");

    /**
     * Pattern to match trailing periods and middle initials in names.
     *
     * This pattern matches:
     *  - A space, followed by a single non-space character (middle initial),
     *    optionally followed by a period, at the end of the string.
     *
     * Breakdown of the regex:
     *   \s           A space
     *   [^\s]        Any single non-space character (middle initial)
     *   \.?          Optional period
     *   $            End of string
     */
    private static final Pattern TRAILING_PERIOD_AND_INITIAL_PATTERN = Pattern.compile("\\s[^\\s]\\.?$");

    public FirstNameAttribute() {
        super(List.of(
                new NotInValidator(
                        AttributeUtilities.COMMON_PLACEHOLDER_NAMES)));
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

    @Override
    public String normalize(String value) {
        String normalizedValue = AttributeUtilities.normalizeDiacritics(value);

        // remove common titles and title abbreviations
        String valueWithoutTitle = TITLE_PATTERN.matcher(normalizedValue).replaceAll("");

        // if the title removal doesn't results in an empty string, use the title-less
        // value, otherwise use title as first name
        if (!valueWithoutTitle.isEmpty()) {
            normalizedValue = valueWithoutTitle;
        }

        String valueWithoutSuffix = AttributeUtilities.GENERATIONAL_SUFFIX_PATTERN.matcher(normalizedValue)
                .replaceAll("");

        // if the generational suffix removal doesn't result in an empty string,
        // continue with the value without suffix, otherwise use the value with suffix
        // as first name
        if (!valueWithoutSuffix.isEmpty()) {
            normalizedValue = valueWithoutSuffix;
        }

        // trim trailing periods
        // remove trailing periods and middle initials
        normalizedValue = TRAILING_PERIOD_AND_INITIAL_PATTERN.matcher(normalizedValue).replaceAll("");

        // remove dashes, spaces and other non-alphanumeric characters
        normalizedValue = AttributeUtilities.NON_ALPHABETIC_PATTERN.matcher(normalizedValue).replaceAll("");

        return normalizedValue;
    }
}
